README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Painbow

<!-- badges: start -->
<!-- badges: end -->

Painbow lets you use [XKCD’s “painbow” colormap](https://xkcd.com/2537/)
in ggplot.

[XKCD](https://xkcd.com/2537/) implied that this colormap is terrible,
and even called it a “painbow”. However, these examples show that with
certain tasks and data, this colormap outperforms even some of the most
commonly cited “good” colormaps like viridis.

<img src="https://imgs.xkcd.com/comics/painbow_award.png" style="width:48.0%" />

Here’s a reproduction in ggplot using some custom theming:  
<img src="man/figures/README-unnamed-chunk-2-1.png" width="48%" />

## Installation

You can install the latest development version:

    install.packages("devtools")
    devtools::install_github("steveharoz/painbow")

## Examples

Setup:

``` r
library(tidyverse)
library(painbow)
library(patchwork) # combine multiple graphs
```

### A simple example for the scale:

``` r
ggplot(faithfuld) +
  aes(waiting, eruptions, fill = density) +
  geom_raster(interpolate = TRUE) +
  scale_fill_painbow() +
  labs(title = "Can you find the most dense region?") +
  theme_bw(18)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" />

### The dataset from the comic

The dataset is `painbow_data`. It was made using the comic’s image and a
scripted lookup table.

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

### Painbow may help you find outliers

Here is a 2D field with a regular pattern and a deviation. Can you find
it? Painbow makes a task easier compared with commonly touted “good”
colormaps.

``` r
##### 2D #####

COUNT = 512

data = expand.grid(
  x = 1:COUNT,
  y = 1:COUNT) %>% 
  mutate(z = sin(x/16) + cos(y/16)) %>% 
  mutate(znoise = z + dnorm(sqrt((x-0.75*COUNT)^2 + (y-0.33*COUNT)^2)/COUNT*20))

ggplot(data) +
  aes(x=x, y=y, fill=znoise) + 
  geom_raster() +
  labs(title = "ggplot default", fill=NULL) + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x=x, y=y, fill=znoise) + 
  geom_raster() +
  scale_fill_viridis_c() +
  labs(title = "Viridis", fill=NULL) + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x=x, y=y, fill=znoise) + 
  geom_raster() +
  scale_fill_painbow() +
  labs(title = "XKCD's colormap", fill=NULL) + 
  theme_void(15) + theme(legend.text = element_blank()) +
patchwork::plot_annotation(
  title = "Three colormaps. Same data. Can you spot the weird region?", 
  theme = theme(text = element_text(size=20)))
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

### Painbow can help you spot a subtle pattern among data with high dynamic range

``` r
######## 1D #########

COUNT = 1024

data = tibble(
  x = 1:COUNT,
  y = x/COUNT + sin(x/4)/100
)

ggplot(data) +
  aes(x = x, y=x) + 
  geom_line() +
  labs(title = "y = x") + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=x) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  labs(title = "ggplot default") + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=x) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  scale_fill_viridis_c() +
  labs(title = "viridis") + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=x) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  scale_fill_painbow() +
  labs(title = "painbow") + 
  theme_void(15) + theme(legend.text = element_blank()) +

ggplot(data) +
  aes(x = x, y=y) + 
  geom_line() +
  labs(title = "y = x + sine wave") +
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=y) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  labs(title = "ggplot default") + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=y) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  scale_fill_viridis_c() +
  labs(title = "viridis") + 
  theme_void(15) + theme(legend.text = element_blank()) +
ggplot(data) +
  aes(x = x, y=COUNT/2, fill=y) + 
  geom_tile(width=1, height=COUNT, color=NA) +
  scale_fill_painbow() +
  labs(title = "painbow") +
  theme_void(15) + theme(legend.text = element_blank()) +

patchwork::plot_layout(ncol=4) +
patchwork::plot_annotation(
    title = "Three colormaps. Same data. Can you spot the harmonic?", 
    theme = theme(text = element_text(size=20)))
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

## Feedback, issues, and contributions

Feedback, suggestions, issues, and contributions are all welcome! Please
file an issue or pull request at
<https://github.com/steveharoz/painbow/issues>

## Citing Painbow

The XKCD comic deserves credit: <https://xkcd.com/2537/>

Please cite this library via:  
Steve Haroz (2021). Painbow. R package version 1.0.0,
<https://github.com/steveharoz/painbow/>.


NEWS.md
# painbow 1.0.1

* Smoother example dataset.
* Documented all exports.

# painbow 1.0.0

* Initial version converted script into package


R/painbow.R
# XKCD's supposedly "bad" colormap https://xkcd.com/2537/
# By Steve Haroz
# License: CC-BY


# define the XKCD "painbow" colormap (extracted from the image)
painbow_colors = c( '#FFFFFF', '#FEFEFF', '#FAFAFC', '#F4F4F7', '#EBECF1', '#E4E4EC', '#DBDAE6', '#D1D1DE', '#C7C7D5', '#BBBDCB', '#AFB3C3', '#A4A7BA', '#999CB0', '#8F91A6', '#87899D', '#7E8193', '#747788', '#6A6D7C', '#626170', '#575765', '#4C4C57', '#41424B', '#37373E', '#2D2C32', '#242328', '#1B1A1E', '#131313', '#0C0C0A', '#060606', '#010102', '#000000', '#000003', '#00000C', '#000219', '#03092A', '#091239', '#111B49', '#162150', '#1C2756', '#222D5C', '#2A3463', '#333B6A', '#3A4370', '#424B77', '#4A527D', '#515A82', '#576085', '#5D6689', '#606C8B', '#60718C', '#5F758A', '#577985', '#4A7E80', '#407D7B', '#367C72', '#287C69', '#1C7C60', '#147C58', '#0C7D50', '#037D47', '#1E7941', '#247139', '#356632', '#54562B', '#714524', '#8D361D', '#AD2516', '#C91610', '#DF0F0C', '#F20605', '#FD0004', '#FE0302', '#FD0400', '#FD0200', '#FC0401', '#FC0202', '#FA0401', '#F80601', '#F80400', '#F50800', '#F30B01', '#F10A01', '#ED0F00', '#E91101', '#E71302', '#E31401', '#E01400', '#DC1900', '#D91D00', '#D41E00', '#D12001', '#CD2200', '#C52A00', '#C32A00', '#C02B01', '#B83201', '#B53301', '#AF3601', '#AA3A01', '#A63D02', '#A04002', '#9A4502', '#964702', '#914A01', '#8B4F01', '#865202', '#825503', '#7D5904', '#775C03', '#716003', '#6D6304', '#666704', '#606B06', '#5D6F07', '#587207', '#527607', '#4E7908', '#4B7C08', '#467F08', '#418309', '#3E860A', '#38890C', '#348D0C', '#2F900C', '#2E930E', '#2A960E', '#269A0E', '#239D0E', '#1FA00F', '#1CA211', '#1EA411', '#1EA712', '#1CA912', '#1BAC14', '#18AE15', '#13B015', '#0DB216', '#0EB417', '#09B715', '#14BA18', '#13BD1B', '#09C11C', '#0FC41E', '#14C721', '#13CA22', '#1CCC25', '#21CF27', '#27D129', '#2CD32A', '#31D62C', '#37D72F', '#41D932', '#47DB35', '#50DD37', '#57DF3A', '#5CE13D', '#65E23F', '#6BE341', '#76E343', '#80E547', '#85E749', '#8BE84D', '#91E950', '#98EB53', '#A1EB56', '#A7ED58', '#ABEE5B', '#B2EE5E', '#BAEE61', '#BEF064', '#C0F167', '#C5F26A', '#CAF26D', '#CFF173', '#D4F17A', '#D6F181', '#D9F186', '#DBF38B', '#DDF293', '#DEF399', '#E0F29F', '#E2F2A8', '#E3F1AF', '#E3F0B5', '#E4EFBC', '#E4EEC1', '#E5EDC8', '#E5ECCD', '#E6EBD2', '#E6EAD8', '#E7E9DD', '#E7E9E1' )


## base graphics palette
painbow.colors = function(n, alpha = FALSE, rev = FALSE) {
  ## n: number of colors
  n = as.integer(n[1L])
  if (n < 1L) return(character())

  ## alpha: logical or numeric alpha transparency (vector) in [0, 1]
  if (identical(alpha, FALSE)) alpha = NULL
  if (!is.null(alpha)) alpha = pmin(pmax(as.numeric(alpha), 0), 1) * 255

  ## interpolate painbow RGB coordinates
  pain = grDevices::col2rgb(painbow_colors)
  x = seq.int(0, 1, length.out = length(painbow_colors))
  xout = seq.int(0, 1, length.out = n)
  pain = grDevices::rgb(
    red   = stats::approx(x, pain[1L, ], xout = xout)$y,
    green = stats::approx(x, pain[2L, ], xout = xout)$y,
    blue  = stats::approx(x, pain[3L, ], xout = xout)$y,
    alpha = alpha,
    maxColorValue = 255)

  ## rev: reverse order?
  if (rev) rev(pain) else pain
}


## ggplot2 scales
scale_fill_painbow = function(...) {
  ggplot2::scale_fill_gradientn(
    colors = painbow_colors,
    values = seq(0, 1, length.out = length(painbow_colors)),
    ...
  )}

scale_color_painbow = function(...) {
  ggplot2::scale_color_gradientn(
    colors = painbow_colors,
    values = seq(0, 1, length.out = length(painbow_colors)),
    ...
  )}

scale_colour_painbow = scale_color_painbow


extra content/extract dataset.R
# extract a dataset from the XKCD image

library("painbow")
library("png")
library("tidyverse")
library("patchwork")


#### get the image data

data_image = readPNG("extra content/dataset.png")
# convert to a data frame
data_image = reshape2::melt(data_image)
data_image = data_image %>% pivot_wider(names_from = Var3, values_from = value)
names(data_image) = c('x', 'y', 'r', 'g', 'b')
data_image = data_image %>% mutate( color = rgb(r, g, b) )

# scale color channels
data_image = data_image %>% mutate(
  r = r*255,
  g = g*255,
  b = b*255
)

# flip orientation
data_image = data_image %>% mutate(
  temp = x,
  x = y,
  y = -temp + max(temp) + 1) %>%
  select(-temp)


# check
ggplot(data_image) +
  aes(x, y, fill = color) +
  geom_raster() +
  scale_fill_identity() +
  theme_classic() +
  coord_equal()



#### generate lookup table for xkcd colors

# example use: interpolate_2_colors("#FFFFFF", "#000000", 0.5)
interpolate_2_colors = function(color1, color2, interpolation) {
  color = colorRamp(c(color1, color2))(interpolation)
  rgb(color[,1]/255, color[,2]/255, color[,3]/255)
}

# keep this high to make sure no values are skipped
interpolation_steps = 64

painbow_colors_lerped = tibble(color = painbow_colors) %>%
  mutate(next_color = lead(color)) %>%
  mutate(index = 1:n()) %>%
  filter(!is.na(next_color)) %>%
  group_by(color, next_color, index) %>%
  # for each row, add additional rows with an interpolation value
  expand(interpolation = seq(0, 1 - 1/interpolation_steps, length.out=interpolation_steps)) %>%
  rowwise() %>%
  mutate(interpolated_color = interpolate_2_colors(color, next_color, interpolation)) %>%
  ungroup() %>%
  # sort and give it a new index
  arrange(index, interpolation) %>%
  mutate(index = 1:n()) %>%
  pull(interpolated_color)

# convert colors to rgb [0-1]
painbow_colors_lerped =
  col2rgb(painbow_colors_lerped) %>%
  t() %>%
  as_tibble() %>%
  rowid_to_column("index") %>%
  mutate(hex = painbow_colors_lerped)

# de-duplicate for faster lookup
painbow_colors_lerped = painbow_colors_lerped %>%
  rowwise() %>%
  mutate(first_instance = which(painbow_colors_lerped$hex == hex)[1]) %>%
  mutate(spread = index - first_instance) %>%
  filter(spread == 0)




#### Convert the image to xkcd colormap index

# 3D distance
calculate_distance = function(x1, y1, z1, x2, y2, z2) {
  sqrt(
    (x1-x2)^2 +
    (y1-y2)^2 +
    (z1-z2)^2
  )
}

# find the closest color
# PARALLELIZED!!!
library(doParallel)
library(itertools)

# for each pixel, compare distance in colorspace with each painbow color
compute = function(data_chunk) {
  for (p in 1:nrow(data_chunk)) {
    if (p %% 100 == 0) print(round(100 * p/nrow(data_chunk), 2))

    pixel = data_chunk[p,]

    closest_distance = Inf
    closest_value = NA
    for (i in 1:nrow(painbow_colors_lerped)) {
      color = painbow_colors_lerped[i,]
      distance = calculate_distance(pixel$r, pixel$g, pixel$b, color$red, color$green, color$blue)
      if (distance < closest_distance) {
        closest_distance = distance
        closest_value = color$index
      }
    }
    data_chunk$value[p] = closest_value
  }
  return(data_chunk)
}

# Create a cluster with 1 fewer cores than are available. Adjust as necessary
mycluster = makeCluster(detectCores() - 1) # 16-1 cores
registerDoParallel(mycluster)
chunksize = ceiling(nrow(data_image)/getDoParWorkers())
# foreach
painbow_data = foreach(
        chunked_data=iter(data_image, by='row', chunksize=chunksize),
        .combine=rbind,
        .export=c('calculate_distance', 'painbow_colors_lerped', 'compute')
        ) %dopar% {
          compute(chunked_data)
        }
# stop
stopCluster(mycluster)

# scale values from 0 to 1
painbow_data = painbow_data %>% mutate(
  value = value / interpolation_steps / length(painbow_colors) * 120
)

# cleanup
painbow_data = painbow_data %>% select(x, y, value)

# test
ggplot(painbow_data) +
  aes(x, y, fill = value) +
  geom_raster(interpolate = TRUE) +
  scale_fill_gradientn(colors=c("white", "black", "white", "black", "white", "black", "white", "black")) +
  theme_classic() +
ggplot(painbow_data) +
  aes(x, y, fill = value) +
  geom_raster(interpolate = TRUE) +
  scale_fill_painbow() +
  theme_classic()


# #### expand the dataset for smoother rendering
#
# # smooth a dataset for prettier graphs
# # data must have x, y, and value
# interpolate_dataset = function(data, scale=10) {
#   # set up the indices and interpolation levels for lookup
#   expanded = expand.grid(
#     x = (1:(max(data$x) * scale)) / scale,
#     y = (1:(max(data$y) * scale)) / scale) %>%
#     mutate(
#       x_prop = x %% 1.0,
#       y_prop = y %% 1.0,
#       x1 = floor(x) + 1,
#       y1 = floor(y) + 1,
#       x2 = ceiling(x+0.001) + 1,
#       y2 = ceiling(y+0.001) + 1
#     ) %>%
#     filter(x2 <= max(data$x), y2 <= max(data$y))
#
#   # make the original data a nicely indexable matrix
#   painbow_matrix = data %>% arrange(y, x) %>% pivot_wider(id_cols = x, names_from = y) %>% select(-x) %>% as.matrix()
#
#   # 2D lerp
#   lerp2D = function(d) {
#     #print(d)
#         v11 = painbow_matrix[d$x1, d$y1]
#         v12 = painbow_matrix[d$x1, d$y2]
#         v21 = painbow_matrix[d$x2, d$y1]
#         v22 = painbow_matrix[d$x2, d$y2]
#         lo = v11 * (1-d$x_prop) + v21 * d$x_prop
#         hi = v12 * (1-d$x_prop) + v22 * d$x_prop
#         return(lo * (1-d$y_prop) + hi * d$y_prop)
#       }
#
#   # lookup based on the indices and interpolate the values
#   expanded = expanded %>%
#     rowwise() %>%
#     mutate( value = lerp2D(cur_data()) ) %>%
#     ungroup()
#
#   # cleanup
#   expanded = expanded %>% select(x, y, value)
#
#   return(expanded)
# }
#
# # test
# painbow_data = painbow_data %>% interpolate_dataset(scale = 10)
# ggplot(temp) +
#   aes(x, y, fill = value) +
#   geom_raster() +
#   scale_fill_painbow() +
#   theme_classic()

usethis::use_data(painbow_data, overwrite = TRUE)
